### PR TITLE
Make boundary forcing optional to support global forecasting

### PR DESCRIPTION
**Pull request by nehamanoj1105**
Description

This PR makes boundary forcing optional in ARModel, enabling support for global forecasting where no boundary conditions are present.

As discussed in Issue #63, global forecasting can be viewed as a special case of LAM without boundary overwriting. Previously, boundary masking was always applied during autoregressive rollout, which prevented clean support for global datasets.

Changes

Made boundary_mask optional during initialisation.

Introduced self.use_boundary_forcing flag.

Updated unroll_prediction() to:

Apply boundary overwrite only when the boundary mask exists.

Skip overwrite entirely for global forecasting.

Ensured loss and metrics behaviour remains unchanged:

LAM: loss computed over interior only.

Global: loss computed over the full grid.

### CODE REVIEW

**Review by InFiNiTy0639 — COMMENTED** [review]
I noticed you've added handling for when da_boundary_mask is None and introduced the use_boundary_forcing flag.  Could you briefly clarify the downstream use case motivating this change? Is this specifically to support global datasets, or are there other models/datasets we're planning to introduce that don't require boundary forcing?

**Comment by nehamanoj1105** [comment]
The primary motivation is to support global forecasting as discussed in Issue #63. In the global setting, the domain covers the entire sphere and therefore has no physical boundary where true states should be injected during rollout. In that case, boundary overwriting is not applicable, and the model should evolve autoregressively over the full grid.

At the moment, the intended downstream use case is global datasets (e.g., ERA5-style global forecasting). The change is structured generically so that any future dataset without boundary conditions can also use the same logic without modification.

**Comment by esmacimsit** [comment]
One thing I wanted to double-check:
since self.boundary_mask can now be None, are there any other methods in ARModel (or inherited classes) that access self.boundary_mask directly without guarding on self.use_boundary_forcing?

I checked unroll_prediction() which is properly protected, but I just want to ensure there are no downstream assumptions that boundary_mask always exists.

Also, do you think adding a minimal unit test covering the da_boundary_mask=None case would be useful to avoid regressions?

**Comment by nehamanoj1105** [comment]
I reviewed the rest of ARModel, and boundary_mask is only used inside unroll_prediction(), which is now guarded via self.use_boundary_forcing. Other components (loss, metrics, masking) rely on interior_mask, which is always defined, including in the global case where it defaults to all ones.

So there should not be any downstream assumptions that boundary_mask is always present.

Regarding testing, yes, I agree that adding a minimal unit test for the da_boundary_mask=None case would be useful to prevent regressions. I can add a small test that initialises ARModel with boundary_mask=None and verifies that unroll_prediction() runs without applying overwrite logic.

File: neural_lam/models/ar_model.py [hunk 1]
```diff
@@ -229,50 +238,55 @@ def predict_step(self, prev_state, prev_prev_state, forcing):
...
-        )  # (B, pred_steps, num_grid_nodes, d_f)
+        prediction = torch.stack(prediction_list, dim=1)
+
         if self.output_std:
-            pred_std = torch.stack(
-                pred_std_list, dim=1
-            )  # (B, pred_steps, num_grid_nodes, d_f)
+            pred_std = torch.stack(pred_std_list, dim=1)
         else:
-            pred_std = self.per_var_std  # (d_f,)
```
Comment (joeloskarsson) [inline neural_lam/models/ar_model.py:275]: What is the motivation for these changes? They seem unrelated

File: neural_lam/models/ar_model.py [hunk 2]
```diff
@@ -124,17 +124,26 @@ def __init__(
...
-        ).unsqueeze(
-            1
-        )  # add feature dim
-
-        self.register_buffer("boundary_mask", boundary_mask, persistent=False)
-        # Pre-compute interior mask for use in loss function
-        self.register_buffer(
-            "interior_mask", 1.0 - self.boundary_mask, persistent=False
-        )  # (num_grid_nodes, 1), 1 for non-border
+        if da_boundary_mask is not None:
```
Comment (joeloskarsson) [inline neural_lam/models/ar_model.py:127]: This might seem like a reasonable check to determine if we should use boundary forcing or not, however right now there is not DataStore that does not have a boundary mask, so we could not actually run the non-boundary-forcing case here. How is the idea that this would be used?

**Review by joeloskarsson — CHANGES_REQUESTED** [review]
Thanks for starting to look at this. I mostly just want some clarifications here, before diving deeper.

**Comment by sadamov** [comment]
@nehamanoj1105 is there still interest in working on this PR from your side or could someone else take over here?

**Comment by kshirajahere** [comment]
@sadamov can I volunteer to work on this?

**Comment by sadamov** [comment]
@kshirajahere I saw that @joeloskarsson and other gsoc-mentors are currently waiting on the proposal review process before assigning issues that are covered by the gsoc projects. It makes sense that the selected applicants would be involved in solving these issues. So we should probably stand by for a bit here.

And most likely any solution to #90 will introduce optional boundary forcing, thereby solving this PR.

**Comment by sadamov** [comment]
Closing this one. The GSoC project on global forecasting (issue #63) did not materialise. Proper boundary handling is now being introduced via #635 (optional boundary datastore), which makes the `use_boundary_forcing` flag here redundant. Thanks @nehamanoj1105 for the contribution.
